text/0000-buildpack-uris.md

```diff
@@ -0,0 +1,128 @@
+
+_Please note that the scheme denotes protocol and not type._
+
+# Motivation
+[motivation]: #motivation
+
+Given that we support a multitude of sources for buildpacks, a certain level of ambiguity has naturally presented
+itself. We currently _try_ to resolve the various sources through a complex set of rules yet the ambiguity to users
+still exists. In config files, this ambiguity is at times resolved by introducing a separate key name to signify
+the source (protocol).
+
+The aim of this RFC is to propose a set of URI schemes that could be used in both CLI arguments and config files
+while solving the ambiguity problem.
+
+A few references to further express the problem:
+
+- [A poll](https://buildpacks.slack.com/archives/C94UJCNV6/p1580920669081100) to understand what users expect when
+providing buildpacks and there is overlap of sources.
+- [Issue](https://github.com/buildpacks/pack/issues/489) to provide additional clarity in logs due to ambiguity. 
+- [A conversation](https://buildpacks.slack.com/archives/CJ6B92ZSB/p1578502454017000) when trying to determine how to
+provide reference to all buildpacks in builder.
+- [Issue](https://github.com/buildpacks/pack/issues/376) where an `image` key was added to differentiate between
+directory and image. 
+ 
+# What it is
+[what-it-is]: #what-it-is
+
+### Scheme
+
+| Short Name | Format | Examples |
+|--- |--- |--- |
+| Filesystem | `file://[<host>]/<path>` | `file:///./my/buildpack.tgz`<br>`file:///home/user/my/buildpack.tgz`
+| URL | `http[s]://<host>/<path>` | `http://example.com/my/buildpack.tgz`<br>`https://example.com/my/buildpack.tgz`  
+| Docker | `docker://[<host>]/<path>[:<tag>]` | `docker:///ubuntu:latest`<br>`docker://gcr.io/distroless/nodejs` 
+| CNB Builder | `cnb:builder://[<id>[@<version>]]` | `cnb:builder://`<br>`cnb:builder://bp.id`<br>`cnb:builder://bp.id@bp.version`
+
+
+### Filesystem
+
+The filesystem scheme is the implementation of [RFC 8089](https://tools.ietf.org/html/rfc8089).
+
+### URL
+
+The URL scheme is the implementation of [RFC 7230](https://tools.ietf.org/html/rfc7230#section-2.7.1).
+
+### Docker
+
+The docker scheme denotes the use of the [Docker HTTP API v2 protocol](https://docs.docker.com/registry/spec/api/).
+
+- `<host>` is optional and defaults to `index.docker.io`.
+- Similar to the `file` scheme, there is a minimal declaration for omitting host by using a simple slash (`/`). 
+eg. `docker:/ubuntu:latest`
+
+### CNB Builder
+
+- `cnb:builder://` - A reference to ALL the buildpacks in the builder.
+- `cnb:builder://bp.id` - A reference to a buildpack with id `bp.id` in the builder. If there is more than one version
+available, a error should be thrown.
+- `cnb:builder://bp.id@bp.version` - A reference to a buildpack with id `bp.id` at the specific version `bp.version`.
+
+# How it Works
+[how-it-works]: #how-it-works
+
+### Usage
+
+Certain `pack` CLI arguments would take URIs supported instead of the current inconsistent format.
```

Comment:
Just to be clear, you mean that the naked `foo/bar` format would be disallowed?

Reply:
Right unless we want to make a very explicit default but I thought we would want to wait until the registry is available.

Reply:
Is there anything preventing us from accepting the naked format, and *always* having it mean one thing (i.e. we always put `cnb:builder://` or whatever in front of it)? That means you only need to add a scheme if you want a local `file://` or whatever.

Reply:
Yes, that's listed under additional considerations: https://github.com/buildpacks/rfcs/blob/buildpack-uris/text/0000-buildpack-uris.md#additional-considerations

Also, it seems like we might need to have the "naked" format be relative URIs. See @ekcasey comment below.